File: app/src/main/java/com/iceteaviet/fastfoodfinder/data/remote/FirebaseClientAuth.java
package com.iceteaviet.fastfoodfinder.data.remote;

import android.support.annotation.NonNull;

import com.google.android.gms.tasks.OnCanceledListener;
import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.firebase.auth.AuthCredential;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.iceteaviet.fastfoodfinder.data.remote.user.model.User;

import io.reactivex.Single;
import io.reactivex.SingleEmitter;
import io.reactivex.SingleOnSubscribe;

/**
 * Created by Genius Doan on 14/07/2017.
 */

public class FirebaseClientAuth implements ClientAuth {
    private FirebaseAuth mAuth;
    private User currentUser;

    public FirebaseClientAuth() {
        mAuth = FirebaseAuth.getInstance();
    }

    @Override
    public String getCurrentUserUid() {
        if (mAuth.getCurrentUser() != null)
            return mAuth.getCurrentUser().getUid();
        return null;
    }

    @Override
    public boolean isSignedIn() {
        return mAuth.getCurrentUser() != null && !mAuth.getCurrentUser().isAnonymous();
    }

    @Override
    public void signOut() {
        mAuth.signOut();
    }

    @Override
    public Single<Boolean> signInWithEmailAndPassword(final String email, final String password) {
        return Single.create(new SingleOnSubscribe<Boolean>() {
            @Override
            public void subscribe(final SingleEmitter<Boolean> emitter) {
                mAuth.signInWithEmailAndPassword(email, password)
                        .addOnSuccessListener(new OnSuccessListener<AuthResult>() {
                            @Override
                            public void onSuccess(AuthResult authResult) {
                                emitter.onSuccess(true);
                            }
                        })
                        .addOnFailureListener(new OnFailureListener() {
                            @Override
                            public void onFailure(@NonNull Exception e) {
                                emitter.onError(e);
                            }
                        })
                        .addOnCanceledListener(new OnCanceledListener() {
                            @Override
                            public void onCanceled() {
                                emitter.onError(new Exception("Cancel"));
                            }
                        });
            }
        });
    }

    @Override
    public Single<FirebaseUser> signInWithCredential(final AuthCredential authCredential) {
        return Single.create(new SingleOnSubscribe<FirebaseUser>() {
            @Override
            public void subscribe(final SingleEmitter<FirebaseUser> emitter) {
                mAuth.signInWithCredential(authCredential)
                        .addOnSuccessListener(new OnSuccessListener<AuthResult>() {
                            @Override
                            public void onSuccess(AuthResult authResult) {
                                emitter.onSuccess(authResult.getUser());
                            }
                        })
                        .addOnFailureListener(new OnFailureListener() {
                            @Override
                            public void onFailure(@NonNull Exception e) {
                                emitter.onError(e);
                            }
                        })
                        .addOnCanceledListener(new OnCanceledListener() {
                            @Override
                            public void onCanceled() {
                                emitter.onError(new Exception("Cancel"));
                            }
                        });
            }
        });
    }

    @Override
    public User getCurrentUser() {
        return currentUser;
    }

    @Override
    public void setCurrentUser(User user) {
        currentUser = user;
    }
}

File: app/src/main/java/com/iceteaviet/fastfoodfinder/data/auth/FirebaseClientAuth.kt
package com.iceteaviet.fastfoodfinder.data.auth

import com.google.firebase.auth.AuthCredential
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import com.iceteaviet.fastfoodfinder.data.remote.user.model.User
import kotlinx.coroutines.tasks.await

class FirebaseClientAuth : ClientAuth {
    private val mAuth: FirebaseAuth

    init {
        mAuth = FirebaseAuth.getInstance()
    }

    override fun getCurrentUserUid(): String {
        val currUser = mAuth.currentUser
        return if (currUser != null) currUser.uid else ""
    }

    override suspend fun signUpWithEmailAndPassword(email: String, password: String): User {
        val result = mAuth.createUserWithEmailAndPassword(email, password).await()
        return convertFirebaseUserToUser(result.user!!)
    }

    override fun isSignedIn(): Boolean {
        val currUser = mAuth.currentUser
        return currUser != null && !currUser.isAnonymous
    }

    override fun signOut() {
        mAuth.signOut()
    }

    override suspend fun signInWithEmailAndPassword(email: String, password: String): User {
        val result = mAuth.signInWithEmailAndPassword(email, password).await()
        return convertFirebaseUserToUser(result.user!!)
    }

    override suspend fun signInWithCredential(authCredential: AuthCredential): User {
        val result = mAuth.signInWithCredential(authCredential).await()
        return convertFirebaseUserToUser(result.user!!)
    }

    private fun convertFirebaseUserToUser(firebaseUser: FirebaseUser): User {
        var photoUrl = ""

        if (firebaseUser.photoUrl != null) {
            photoUrl = firebaseUser.photoUrl.toString()
        }

        var displayName = ""
        firebaseUser.displayName?.let {
            displayName = it
        }

        val email = firebaseUser.email
        if (email != null)
            return User(firebaseUser.uid, displayName, email, photoUrl, ArrayList())
        else
            throw IllegalArgumentException()
    }
}
